src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,141 @@
+            s_localizableInsideMessage,
+            configurable: true);
+
+        internal static readonly DiagnosticDescriptor _outsideDescriptor = CreateDescriptorWithId(
+            IDEDiagnosticIds.MoveMisplacedUsingsDiagnosticId,
+            s_localizableTitle,
+            s_localizableOutsideMessage,
+            configurable: true);
+
+        private readonly ICodingConventionsManager _conventionsManager = CodingConventionsManagerFactory.CreateCodingConventionsManager();
+
+        public MisplacedUsingsDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.MoveMisplacedUsingsDiagnosticId, s_localizableTitle)
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.CompilationUnit);
+            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.NamespaceDeclaration);
+        }
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            var statement = context.Node;
+            var cancellationToken = context.CancellationToken;
+
+            var option = GetPreferredPlacementOption(context);
+            if (option.Value == UsingPlacementPreference.NoPreference)
+            {
+                return;
+            }
+
+            if (statement is CompilationUnitSyntax compilationUnit
+                && option.Value == UsingPlacementPreference.InsideNamespace
+                && !ShouldSurpressDiagnostic(compilationUnit))
+            {
+                ReportDiagnostics(context, _insideDescriptor, compilationUnit.Usings, option);
+            }
+            else if (statement is NamespaceDeclarationSyntax namespaceDeclaration
+                && option.Value == UsingPlacementPreference.OutsideNamespace)
+            {
+                ReportDiagnostics(context, _outsideDescriptor, namespaceDeclaration.Usings, option);
+            }
+        }
+
+        private CodeStyleOption<UsingPlacementPreference> GetPreferredPlacementOption(SyntaxNodeAnalysisContext context)
+        {
+            var statement = context.Node;
+            var cancellationToken = context.CancellationToken;
+
+            // While running in VisualStudio this should always return an OptionSet.
+            var optionSet = context.Options.GetDocumentOptionSetAsync(statement.SyntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet != null)
+            {
+                return optionSet.GetOption(CSharpCodeStyleOptions.PreferredUsingPlacement);
+            }
+
+            // This code path is for tests only. It relies on the test creating a .editorconfig in the current working directory.
+            var filePath = context.SemanticModel.SyntaxTree.FilePath;
+            var fullPath = string.IsNullOrEmpty(Path.GetDirectoryName(filePath))
+                ? Path.Combine(Environment.CurrentDirectory, filePath)
+                : filePath;
```

Comment:
i don't understand this.  NO other features do this, so it feels very wrong to me for this feature to do this.

Reply:
➡️ Requirement for injecting .editorconfig values into the new test framework, prior to full adoption of the upcoming compiler-layer APIs for .editorconfig. We tried hard to avoid this, but all options (aside from waiting for the new APIs) were ugly.

Reply:
This should be encapsulated somewhere so that this can happen without each individual feature needing to have hardcoded in knowledge of this sort of hting.  That way the brittleness is isolated to a single location, isntead of virally spreading into every feature.